subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/WithSideEffectProviderTest.groovy

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider
+
+import spock.lang.Specification
+
+import java.util.concurrent.atomic.AtomicInteger
+
+
+class WithSideEffectProviderTest extends Specification {
```

Comment:
There is `ProviderSpec` which has some tests which should work on all providers. I think it would be good to add some tests for the side effects there as well. And to the corresponding subclasses if necessary, for example for the collection providers.